streamdeck.go

```diff
@@ -287,47 +212,155 @@ func (sd *StreamDeck) FillColor(btnIndex, r, g, b int) error {
-	color := color.RGBA{uint8(r), uint8(g), uint8(b), 0}
+	img := image.NewRGBA(image.Rect(0, 0, sd.config.ButtonSize, sd.config.ButtonSize))
+	color := color.RGBA{uint8(r), uint8(g), uint8(b), 1}
 	draw.Draw(img, img.Bounds(), image.NewUniform(color), image.Point{0, 0}, draw.Src)
 
 	return sd.FillImage(btnIndex, img)
 }
 
+func (sd *StreamDeck) encodeImage(img image.Image) ([]byte, error) {
+	if sd.config.ImageRotate {
+		newImage := image.NewRGBA(img.Bounds())
+		for x := 0; x < sd.config.ButtonSize; x++ {
+			for y := 0; y < sd.config.ButtonSize; y++ {
+				newImage.Set(x, y, img.At(sd.config.ButtonSize-x, sd.config.ButtonSize-y))
+			}
+		}
+		img = newImage
+	}
+
+	if sd.config.ImageFormat == "bmp" {
+		return encodeBMP(sd.config, img)
+	}
+
+	if sd.config.ImageFormat == "jpg" {
+		buf := bytes.Buffer{}
+		err := jpeg.Encode(&buf, img, nil)
+		if err != nil {
+			return nil, err
+		}
+		return buf.Bytes(), nil
+	}
+
+	return nil, fmt.Errorf("unknown image format [%s]", sd.config.ImageFormat)
+
+}
+
+func encodeBMP(c Config, img image.Image) ([]byte, error) {
+	imgBuf := []byte{
+		'\x42', '\x4D', '\xF6', '\x3C', '\x00', '\x00', '\x00', '\x00',
+		'\x00', '\x00', '\x36', '\x00', '\x00', '\x00', '\x28', '\x00',
+		'\x00', '\x00', '\x48', '\x00', '\x00', '\x00', '\x48', '\x00',
+		'\x00', '\x00', '\x01', '\x00', '\x18', '\x00', '\x00', '\x00',
+		'\x00', '\x00', '\xC0', '\x3C', '\x00', '\x00', '\xC4', '\x0E',
+		'\x00', '\x00', '\xC4', '\x0E', '\x00', '\x00', '\x00', '\x00',
+		'\x00', '\x00', '\x00', '\x00', '\x00', '\x00'}
+
+	for row := 0; row < c.ButtonSize; row++ {
+		for line := c.ButtonSize - 1; line >= 0; line-- {
+			r, g, b, _ := img.At(line, row).RGBA()
+			imgBuf = append(imgBuf, byte(b), byte(g), byte(r))
+		}
+	}
+	return imgBuf, nil
+}
+
 // FillImage fills the given key with an image. For best performance, provide
 // the image in the size of 72x72 pixels. Otherwise it will be automatically
 // resized.
 func (sd *StreamDeck) FillImage(btnIndex int, img image.Image) error {
-	if err := checkValidKeyIndex(btnIndex); err != nil {
+	if err := sd.checkValidKeyIndex(btnIndex); err != nil {
 		return err
 	}
 
 	// if necessary, rescale the picture
 	rect := img.Bounds()
-	if rect.Dx() != ButtonSize {
-		img = resize(img, ButtonSize, ButtonSize)
+	if rect.Dx() != sd.config.ButtonSize {
+		img = resize(img, sd.config.ButtonSize, sd.config.ButtonSize)
 	}
 
-	imgBuf := make([]byte, 0, ButtonSize*ButtonSize*3)
+	imgBuf, err := sd.encodeImage(img)
+	if err != nil {
+		return err
+	}
 
-	for row := 0; row < ButtonSize; row++ {
-		for line := ButtonSize - 1; line >= 0; line-- {
-			r, g, b, _ := img.At(line, row).RGBA()
-			imgBuf = append(imgBuf, byte(r), byte(b), byte(g))
+	sd.lock.Lock()
+	defer sd.lock.Unlock()
+
+	if sd.config.ImageFormat == "bmp" {
+		err := sd.sendOriginalSingleMsgInLock(btnIndex, 1, imgBuf[0:7776])
+		if err != nil {
+			return err
 		}
+
+		return sd.sendOriginalSingleMsgInLock(btnIndex, 2, imgBuf[7776:])
 	}
```

Comment:
the split happens after 7803 bytes (2583 pixels resulting in 7749 bytes + 53 bytes of ExtraHeader). 
So [0:7803] and [7803:] would be correct.